### PR TITLE
fix(restore): multiple restore requests should be rejected and proposals should not be submitted

### PR DESCRIPTION
**Pull request by aman-bansal**
Issue: multiple restore requests resulted in multiple restore operation even though requests erred out.

Reason: We have a check to manage number of restore operations running. But the check is happening after submitted the proposals hence failure message but multiple restore operations trigger.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7118)

### CODE REVIEW

**Review by danielmai — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7118)* status: all files reviewed, 1 unresolved discussion (waiting on @aman-bansal, @manishrjain, @MichaelJCompton, and @vvbalaji-dgraph)

___
*[wiki/content/enterprise-features/binary-backups.md, line 431 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-MOIMi_1CJE7skBMpDYw:-MOIMi_26dG2_GKiimaX:bye520w) ([raw file](https://github.com/dgraph-io/dgraph/blob/bde9060e84543081284d101d4f7c11dada8a968e/wiki/content/enterprise-features/binary-backups.md#L431)):*
> ```
> Only restore once. If you run same restore on a different node at the same time,
> then that request will also be accepted.
> ```
When running a live restore, you only need to send the request to a single Alpha. If you run the same restore on a different Alpha at the same time, then another restore operation will be started.

**Review by danielmai — COMMENTED** [review]
Update the doc and then :lgtm:

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-:-MOIN2Ur4EhWDpjPkUo-:bn50en1)* status: all files reviewed, 1 unresolved discussion (waiting on @aman-bansal, @manishrjain, @MichaelJCompton, and @vvbalaji-dgraph)

File: worker/online_restore_ee.go [hunk 1]
```diff
@@ -62,10 +62,14 @@ func ProcessRestoreRequest(ctx context.Context, req *pb.RestoreRequest) (int, er
 	if err := VerifyBackup(req, &creds, currentGroups); err != nil {
 		return 0, errors.Wrapf(err, "failed to verify backup")
 	}
-
 	if err := FillRestoreCredentials(req.Location, req); err != nil {
 		return 0, errors.Wrapf(err, "cannot fill restore proposal with the right credentials")
 	}
+	restoreId, err := rt.Add()
+	if err != nil {
+		return 0, errors.Wrapf(err, "cannot assign ID to restore operation")
```
Comment (manishrjain) [inline worker/online_restore_ee.go:70]: cannot assign ID to restore operation, because of?

**Comment by manishrjain** [comment]
Approved. Just one comment.

File: worker/online_restore_ee.go (hunk 1 above)
Comment (aman-bansal) [reply worker/online_restore_ee.go:70]: wrapf will take care of that. rt.Add() returns multiple kind of errors.  wrapf will combine message to make it more meaningfull. This is the interface it implements. `func (w *withMessage) Error() string { return w.msg + ": " + w.cause.Error() }` Adding something in errors will mess up with the api response. But yes we can add some logging if needed. But seems redundant here.

**Review by manishrjain — CHANGES_REQUESTED** [review]
Hey @aman-bansal , were you going to remove the restroreTracker too? Or, is that a different PR?

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-:-MORsmXT5mOT55605fdI:b-3y6j8o)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @aman-bansal, @danielmai, @MichaelJCompton, and @vvbalaji-dgraph)

___
*[worker/draft.go, line 226 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-MORsPjK27liw3fTkyu7:-MORsPjK27liw3fTkyu8:b-99kilc) ([raw file](https://github.com/dgraph-io/dgraph/blob/240a23ee268413643bfa72b82b39d3439e3bd816/worker/draft.go#L226)):*
> ```Go
> }
> 
> func isTaskRunning(operation op) bool {
> ```

GetOngoingTasks gives you the same information. Use that.

**Review by aman-bansal — COMMENTED** [review]
It will be a separate PR. It will make one endpoint operation `restoreStatus` redundant. I will cherrypick this to 20.11 and restore tracker removal being breaking change, will be released later.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-:-MOVDTR-1OOulkTUkDIF:b6hlzl6)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @danielmai, @manishrjain, @MichaelJCompton, and @vvbalaji-dgraph)

___
*[wiki/content/enterprise-features/binary-backups.md, line 431 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-MOIMi_1CJE7skBMpDYw:-MOVECBwBUyrUB_1a6Xp:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/bde9060e84543081284d101d4f7c11dada8a968e/wiki/content/enterprise-features/binary-backups.md#L431)):*
<details><summary><i>Previously, danielmai (Daniel Mai) wrote…</i></summary><blockquote>

> ```
> Only restore once. If you run same restore on a different node at the same time,
> then that request will also be accepted.
> ```
When running a live restore, you only need to send the request to a single Alpha. If you run the same restore on a different Alpha at the same time, then another restore operation will be started.
</blockquote></details>

Done.
___
*[worker/draft.go, line 226 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-MORsPjK27liw3fTkyu7:-MOVEEoaEHpt4ifdasd3:b-ubvvvy) ([raw file](https://github.com/dgraph-io/dgraph/blob/240a23ee268413643bfa72b82b39d3439e3bd816/worker/draft.go#L226)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

GetOngoingTasks gives you the same information. Use that.
</blockquote></details>

I had to write this logic over loop to check particular op is running or not. I thought being here could be used later. One more thing because of which I was reluctant is GetOngoingTasks return []string. Let me push another thing that uses GetOngoingTasks.

**Review by manishrjain — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7118#-:-MObwUpU43vpP5HxzWsY:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @danielmai, @manishrjain, @MichaelJCompton, and @vvbalaji-dgraph)

**Pull request merged**